File: store-service/internal/packaging/packaging.go
package packaging

type Material string
type Protection string
type Size string
type ShippingMode string

const (
	Wood      Material = "wood"
	Cardboard Material = "cardboard"
	Plastic   Material = "plastic"

	Polystyrene   Protection = "polystyrene"
	BubbleWrap    Protection = "bubble_wrap"
	MoistureBeads Protection = "moisture_beads"

	XLarge Size = "XLarge"
	Large  Size = "Large"
	Medium Size = "Medium"
	Small  Size = "Small"
	XSmall Size = "XSmall"

	Air  ShippingMode = "air"
	Land ShippingMode = "land"
	Sea  ShippingMode = "sea"
)

// --- Strategy pattern -------------------------------------------------------
// Each strategy exposes the material it uses. Concrete strategies are
// unexported — outside the package they're only reachable via Build.

type packagingStrategy interface {
	material() Material
}

type woodPackaging struct{}
type cardboardPackaging struct{}
type plasticPackaging struct{}

func (woodPackaging) material() Material      { return Wood }
func (cardboardPackaging) material() Material { return Cardboard }
func (plasticPackaging) material() Material   { return Plastic }

func strategyForSize(s Size) packagingStrategy {
	switch s {
	case XLarge, Large:
		return woodPackaging{}
	case Medium:
		return cardboardPackaging{}
	case Small, XSmall:
		return plasticPackaging{}
	}
	return nil
}

// --- Decorator pattern ------------------------------------------------------
// Package wraps a selected strategy and adds protections derived from the
// material + shipping mode.

type Package struct {
	material    Material
	protections []Protection
}

func (p Package) Material() Material        { return p.material }
func (p Package) Protections() []Protection { return p.protections }

func protectionsFor(material Material, mode ShippingMode) []Protection {
	switch mode {
	case Air:
		if material == Wood || material == Cardboard {
			return []Protection{Polystyrene}
		}
		return []Protection{BubbleWrap}
	case Land:
		return []Protection{Polystyrene}
	case Sea:
		return []Protection{MoistureBeads, BubbleWrap}
	}
	return nil
}

// Build is the public factory: pick a strategy from size, then decorate it
// with protections derived from the shipping mode.
func Build(size Size, mode ShippingMode) Package {
	s := strategyForSize(size)
	m := s.material()
	return Package{
		material:    m,
		protections: protectionsFor(m, mode),
	}
}

File: backend/src/packaging/packaging.js
// Strategy + Decorator: size → strategy (picks material), then
// (material, shippingMode) → list of protections that decorate the
// resulting Package.

export const MATERIALS = Object.freeze({
  Wood: "wood",
  Cardboard: "cardboard",
  Plastic: "plastic",
});

export const PROTECTIONS = Object.freeze({
  Polystyrene: "polystyrene",
  BubbleWrap: "bubble_wrap",
  MoistureBeads: "moisture_beads",
});

export const SIZES = Object.freeze({
  XLarge: "XLarge",
  Large: "Large",
  Medium: "Medium",
  Small: "Small",
  XSmall: "XSmall",
});

export const SHIPPING_MODES = Object.freeze({
  Air: "air",
  Land: "land",
  Sea: "sea",
});

// Single source of truth for valid shipping modes — callers (validators,
// registries) read from here instead of hardcoding their own switch.
const _shippingModeList = Object.freeze([
  SHIPPING_MODES.Air,
  SHIPPING_MODES.Land,
  SHIPPING_MODES.Sea,
]);

export function shippingModes() {
  return _shippingModeList.slice(); // fresh copy so callers can't mutate
}

export function isValidShippingMode(mode) {
  return _shippingModeList.includes(mode);
}

// --- Strategy: size → material --------------------------------------------
const sizeToMaterial = Object.freeze({
  [SIZES.XLarge]: MATERIALS.Wood,
  [SIZES.Large]: MATERIALS.Wood,
  [SIZES.Medium]: MATERIALS.Cardboard,
  [SIZES.Small]: MATERIALS.Plastic,
  [SIZES.XSmall]: MATERIALS.Plastic,
});

function strategyForSize(size) {
  const material = sizeToMaterial[size];
  if (!material) throw new Error(`unknown size "${size}"`);
  return material;
}

// --- Decorator: (material, mode) → protections ----------------------------
function protectionsFor(material, mode) {
  switch (mode) {
    case SHIPPING_MODES.Air:
      if (material === MATERIALS.Wood || material === MATERIALS.Cardboard) {
        return [PROTECTIONS.Polystyrene];
      }
      return [PROTECTIONS.BubbleWrap];
    case SHIPPING_MODES.Land:
      return [PROTECTIONS.Polystyrene];
    case SHIPPING_MODES.Sea:
      return [PROTECTIONS.MoistureBeads, PROTECTIONS.BubbleWrap];
    default:
      throw new Error(`unknown shipping mode "${mode}"`);
  }
}

// Package is the decorator result — a material + its protections. Frozen
// so callers can pass it around without accidental mutation.
export class Package {
  constructor(material, protections) {
    this.material = material;
    this.protections = Object.freeze(protections.slice());
    Object.freeze(this);
  }
}

// Public factory: size → strategy → decorated package. Throws on unknown
// size or mode so callers bypassing upstream validation can't silently
// produce a partially-populated package.
export function build(size, mode) {
  const material = strategyForSize(size);
  const protections = protectionsFor(material, mode);
  return new Package(material, protections);
}
